File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = 0, kk = 0;
    public void dfs(TreeNode root){
        if(root == null || kk < 0) return;
        dfs(root.left);
        if(--kk == 0){
            ans = root.val;
            --kk;
            return;
        }
        dfs(root.right);
    }
    public int kthSmallest(TreeNode root, int k) {
        kk = k;
        dfs(root);
        return ans;
            
    }
}

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int counter = 0, ans;
    void inorder(TreeNode* root){
        if(!root) return;
        inorder(root->left);
        if(--counter == 0)
            ans = root->val;
        inorder(root->right);
    }
    int kthSmallest(TreeNode* root, int k) {
        counter = k;
        inorder(root);
        return ans;
    }
};